### PR TITLE
[partition] Allow specifying unencrypted partitions when encryption is used

### PR DESCRIPTION
**Pull request by ArrayBolt3**
Implements feature request https://github.com/calamares/calamares/issues/1938 sorta. There's no user-visible checkbox for encrypting or not encrypting `/boot`, instead there is an additional `noEncrypt` option added to the `partitionLayout` section of `partition.conf`, allowing a distro to specify that a certain partition should remain unencrypted even when encryption is enabled.

### CODE REVIEW

File: src/modules/partition/core/PartitionLayout.h
Comment (ArrayBolt3) [inline src/modules/partition/core/PartitionLayout.h:1]: gah, Vim struck again and did a weird mix of spaces plus tabs. Fixing...

Comment (ArrayBolt3) [reply src/modules/partition/core/PartitionLayout.h:1]: Fixed.

**Comment by ArrayBolt3** [comment]
I'm realizing I didn't make this work right with ZFS partitions. I probably should go through and fix that.

File: src/modules/partition/core/PartitionLayout.cpp [hunk 1]
```diff
@@ -116,6 +118,7 @@ PartitionLayout::init( FileSystem::Type defaultFsType, const QVariantList& confi
                           Calamares::getUnsignedInteger( pentry, "attributes", 0 ),
                           Calamares::getString( pentry, "mountPoint" ),
                           Calamares::getString( pentry, "filesystem", "unformatted" ),
+                          Calamares::getBool( pentry, "noEncrypt", true ),
```
Comment (dalto8) [inline src/modules/partition/core/PartitionLayout.cpp]: This should be `false` by default to mirror the prior behavior.

**Comment by ArrayBolt3** [comment]
alright, that should work with ZFS encryption too. I did **not** test whether this results in unencrypted ZFS or not yet, but  I believe it will work as the logic is very straightforward.

File: src/modules/partition/partition.conf [hunk 2]
```diff
@@ -293,6 +295,7 @@ defaultFileSystemType:  "ext4"
 #       - if "unknown" (or an unknown FS name, like "elephant") then the
 #         default filesystem type, or the user's choice, will be applied instead
 #         of "unknown" (e.g. the user might pick ext4, or xfs).
+#   - noEncrypt: whether this partition is exempt from encryption if enabled (optional parameter)
```
Comment (dalto8) [inline src/modules/partition/partition.conf]: We should mention the default state.

File: src/modules/partition/core/PartitionLayout.cpp [hunk 3]
```diff
@@ -296,7 +299,7 @@ PartitionLayout::createPartitions( Device* dev,
         Partition* part = nullptr;
 
         // Encryption for zfs is handled in the zfs module
-        if ( luksPassphrase.isEmpty() || correctFS( entry.partFileSystem ) == FileSystem::Zfs )
+        if ( luksPassphrase.isEmpty() || correctFS( entry.partFileSystem ) == FileSystem::Zfs || entry.partNoEncrypt )
```
Comment (dalto8) [inline src/modules/partition/core/PartitionLayout.cpp:302]: We need to update the comment above this to reflect these changes.

**Review by dalto8 — CHANGES_REQUESTED** [review]

File: src/modules/partition/core/PartitionLayout.cpp (hunk 1 above)
Comment (ArrayBolt3) [reply src/modules/partition/core/PartitionLayout.cpp]: Good catch, that was an accident. Fixing now.

**Comment by ArrayBolt3** [comment]
Got all changes implemented. (The default variable thing was a mistake since I went from implementing this as an `isEncryptable` option that defaulted to true, to a `noEncrypt` option that defaulted to false, and then forgot to change the default.)

**Comment by dalto8** [comment]
I will take a look at it more depth tomorrow.  If memory serves, that isn't the only place we do the partition layout.(I could be wrong though)

The partition module is quite a mess.

**Comment by ArrayBolt3** [comment]
Thanks!

I did a quick look through and the only other place I saw where
encrypted partitions are made is in PartitionActions.cpp, where an
(optionally encrypted) swap partition is made. I don't particularly
mind if the swap partition is *always* encrypted when encryption is
used - unencrypted swap breaks the security of FDE pretty much
entirely (at best it's a gaping security hole), and if someone's that
determined to configure their system in a broken way, they can use the
manual partitioning screen.

In case it's helpful, I did patch Ubuntu's Calamares locally with (the
earlier version of) these changes and successfully used it to make an
encrypted Kubuntu installation with unencrypted /boot without doing
manual partitioning.

**Review by dalto8 — APPROVED** [review]

**Review by dalto8 — APPROVED** [review]

**Pull request merged**